hive-metastore/src/main/java/org/apache/iceberg/hive/HiveSchemaUtil.java

```diff
@@ -75,7 +83,42 @@ public static Schema convert(List<FieldSchema> fieldSchemas, boolean autoConvert
       typeInfos.add(TypeInfoUtils.getTypeInfoFromTypeString(col.getType()));
       comments.add(col.getComment());
     }
-    return HiveSchemaConverter.convert(names, typeInfos, comments, autoConvert);
+    Schema schema = HiveSchemaConverter.convert(names, typeInfos, comments, autoConvert);
+    return rebuildSchemaWithIdentifierFieldIds(schema, identifierFieldNames);
+  }
+
+  /**
+   * Rebuild a schema with given schema and identifierFieldNames
+   * @param schema The origin schema.
+   * @param identifierFieldNames The identifierFieldNames.
+   * @return New schema with IdentifierFieldIds.
+   */
+  public static Schema rebuildSchemaWithIdentifierFieldIds(Schema schema, Set<String> identifierFieldNames) {
+    // Identifier fields in nested field are not supported, so we just check the first level columns.
+    Map<String, Types.NestedField> columnsMap = schema.columns().stream()
+            .collect(Collectors.toMap(Types.NestedField::name, field -> field));
+    Set<Integer> identifierFieldIds = identifierFieldNames.stream()
+            .map(name -> {
+              Types.NestedField field = columnsMap.get(name);
+              if (field == null) {
+                // Does not exist or in nested field.
+                throw new IllegalArgumentException(
+                        String.format("Cannot add field `%s` as an identifier field: " +
+                                "must not in nested field and exist", name));
```

Comment:
I am not a native English speaker, but maybe this is better:
```
Cannot set field `%s` as an identifier field: the field must exist on the root level
```

Or something better 😄